vllm/model_executor/layers/fused_moe/flashinfer_cutlass_moe.py

```diff
@@ -0,0 +1,254 @@
+            f"FlashInferExperts disabled: w1/w2 must be torch.uint8 (got w1={w1.dtype}, w2={w2.dtype}), "
+            f"hidden_states must be float32, float16, or bfloat16 (got {hidden_states.dtype})."
+        )
+        return False
+    return True
+
+
+class FlashInferExperts(mk.FusedMoEPermuteExpertsUnpermute):
+
+    def __init__(self):
+        super().__init__()
+
+    def supports_chunking(self) -> bool:
+        return True
+
+    def workspace_shapes(
+        self, a: torch.Tensor, aq: torch.Tensor, M: int, N: int, K: int,
+        topk: int, global_num_experts: int, local_num_experts: int
+    ) -> tuple[tuple[int, ...], tuple[int, ...], tuple[int, ...], torch.dtype]:
+        # We use global_num_experts due to how moe_align_block_size handles
+        # expert_maps.
+        # TODO(shuw): Just 0 Here
+        num_experts = global_num_experts
+        block_m = self.block_shape[0]
+        M_sum = (M * topk) + num_experts * (block_m - 1)
+        M_sum = round_up(M_sum, block_m)
+        workspace1 = (M_sum, max(N * 2, K))
+        workspace2 = (M_sum, max(N, K))
+        output = (M * topk, K)
+        return (workspace1, workspace2, output, a.dtype)
+
+    def apply(
+        self,
+        hidden_states: torch.Tensor,
+        w1: torch.Tensor,
+        w2: torch.Tensor,
+        topk_ids: torch.Tensor,
+        topk_weights: torch.Tensor,
+        activation: str,
+        global_num_experts: int,
+        w1_scale: Optional[torch.Tensor],
+        w2_scale: Optional[torch.Tensor],
+        w1_zp: Optional[torch.Tensor],
+        w2_zp: Optional[torch.Tensor],
+        a1_scale: Optional[torch.Tensor],
+        a2_scale: Optional[torch.Tensor],
+        expert_num_tokens: Optional[torch.Tensor],
+        g1_alphas: torch.Tensor,
+        g2_alphas: torch.Tensor,
+        input_sf: torch.Tensor,
+        out_dtype: torch.dtype,
+        ep_rank: Optional[int] = 0,
+        ep_size: Optional[int] = 1,
+        tp_rank: Optional[int] = 0,
+        tp_size: Optional[int] = 1,
+    ):
+        # Flashinfer CUTLASS kernel takes scalar global scales,
+        # min because inv_scale.
+
+        quant_scales = [
+            torch.min(a1_scale),
+            w1_scale.view(torch.int32),
+            g1_alphas,
+            torch.min(a2_scale),
+            w2_scale.view(torch.int32),
+            g2_alphas,
+        ]
+        # Flashinfer CUTLASS moe takes in activations in BF16/Half/nvfp4.
+        return cutlass_fused_moe(
+            hidden_states,
+            topk_ids.to(torch.int),
+            topk_weights,
+            # FlashInfer API requires weight to be long for nvfp4
+            w1.view(torch.long),
+            w2.view(torch.long),
+            out_dtype,
+            quant_scales=quant_scales,
+            input_sf=input_sf,
+            ep_size=ep_size,
+            ep_rank=ep_rank,
+            tp_size=tp_size,
+            tp_rank=tp_rank,
+        )[0]
+
+
+class FlashInferCutlassKernels(mk.FusedMoEModularKernel):
```

Comment:
The `FusedMoEModularKernel` isn't intended to be subclassed.  It should be generic enough so that any prepare/finalze + experts should work.  If there are features of these new kernels that don't quite fit, we should update the abstract interfaces or make the base `FusedMoEModularKernel` handle them properly.